### PR TITLE
Decoupled transport-api and its implementation

### PR DESCRIPTION
**Pull request by segabriel**

### CODE REVIEW

File: benchmark/src/main/java/io/scalecube/cluster/gossip/ProtostuffGossipBenchmark.java [hunk 1]
```diff
@@ -1,9 +1,9 @@
 package io.scalecube.cluster.gossip;
 
 import io.scalecube.cluster.Member;
-import io.scalecube.transport.Address;
-import io.scalecube.transport.Message;
-import io.scalecube.transport.MessageCodec;
+import io.scalecube.cluster.transport.api.Address;
+import io.scalecube.cluster.transport.api.Message;
+import io.scalecube.cluster.transport.api.MessageCodec;
```
Comment (artem-v) [inline benchmark/src/main/java/io/scalecube/cluster/gossip/ProtostuffGossipBenchmark.java]: The transition `io.scalecube.transport` -> `io.scalecube.cluster.transport.api` causes a lot of changed files. Let's split it into separate PR.

File: transport-api/src/main/java/io/scalecube/cluster/transport/api/TransportFactory.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package io.scalecube.cluster.transport.api;
+
+@FunctionalInterface
+public interface TransportFactory {
+
+  static Transport defaultTransport(TransportConfig transportConfig) {
```
Comment (artem-v) [inline transport-api/src/main/java/io/scalecube/cluster/transport/api/TransportFactory.java]: Rename `defaultTransport` -> `loadTransport`

File: benchmark/src/main/java/io/scalecube/cluster/gossip/ProtostuffGossipBenchmark.java (hunk 1 above)
Comment (artem-v) [reply benchmark/src/main/java/io/scalecube/cluster/gossip/ProtostuffGossipBenchmark.java]: `io.scalecube.cluster.transport.api` -> `io.scalecube.cluster.transport` ;  it will live in oackage with prefix ` -api `  so it's already clear that this is `api`

File: cluster/src/main/java/io/scalecube/cluster/Cluster.java [hunk 3]
```diff
@@ -95,22 +96,18 @@ static Cluster joinAwait(ClusterConfig config) {
   /**
    * Init cluster instance with the given configuration and join cluster synchronously.
    */
-  static CompletableFuture<Cluster> join(final ClusterConfig config) {
-    return new ClusterImpl(config).join0();
+  static Mono<Cluster> join(final ClusterConfig config) {
```
Comment (artem-v) [inline cluster/src/main/java/io/scalecube/cluster/Cluster.java]: Plz dont substitute `futures` to reactor `publishers`. Let's do this in separate PR.

File: cluster/src/main/java/io/scalecube/cluster/Cluster.java [hunk 4]
```diff
@@ -178,7 +175,7 @@ static Cluster joinAwait(ClusterConfig config) {
    *
    * @return Listenable future which is completed once graceful shutdown is finished.
    */
-  CompletableFuture<Void> shutdown();
+  Mono<Void> shutdown();
```
Comment (artem-v) [inline cluster/src/main/java/io/scalecube/cluster/Cluster.java]: Plz dont substitute `futures` to reactor `publishers`. Let's do this in separate PR.

File: cluster/src/main/java/io/scalecube/cluster/Cluster.java (hunk 3 above)
Comment (ronenhamias) [reply cluster/src/main/java/io/scalecube/cluster/Cluster.java]: i guess if we use rsocket it comes more native to use reactor rather then futures

File: transport-api/pom.xml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      <artifactId>scalecube-testlib</artifactId>
+      <version>${project.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.projectreactor</groupId>
+      <artifactId>reactor-core</artifactId>
+    </dependency>
+
+    <dependency>
```
Comment (ronenhamias) [inline transport-api/pom.xml]: transport api cannot have dependency on netty.

its should represent clear pattern of providing data only.

File: transport-api/src/main/java/io/scalecube/cluster/transport/api/MessageSchema.java [hunk 6]
```diff
@@ -1,7 +1,7 @@
-package io.scalecube.transport;
+package io.scalecube.cluster.transport.api;
 
 import static io.protostuff.LinkedBuffer.MIN_BUFFER_SIZE;
-import static io.scalecube.transport.RecyclableLinkedBuffer.DEFAULT_MAX_CAPACITY;
+import static io.scalecube.cluster.transport.api.RecyclableLinkedBuffer.DEFAULT_MAX_CAPACITY;
```
Comment (ronenhamias) [inline transport-api/src/main/java/io/scalecube/cluster/transport/api/MessageSchema.java]: this seem like netty specifics

File: transport-api/src/main/java/io/scalecube/cluster/transport/ServiceLoaderUtil.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public final class ServiceLoaderUtil {
```
Comment (ronenhamias) [inline transport-api/src/main/java/io/scalecube/cluster/transport/ServiceLoaderUtil.java]: i think we should coniser this utility as part of scalecube utils and not as part of transport
as it appears to be common across projects.

Comment (ronenhamias) [reply transport-api/src/main/java/io/scalecube/cluster/transport/ServiceLoaderUtil.java]: i saw you added it to app utils and i was thinking app utils is more about packaging
maybe we just need new module scalecube-utils....

File: pom.xml [hunk 8]
```diff
@@ -170,6 +173,25 @@
         <version>${netty.version}</version>
       </dependency>
 
+      <!-- RSocket -->
+      <dependency>
```
Comment (ronenhamias) [inline pom.xml]: why adding rsocket as part of this pull request?

**Comment by artem-v** [comment]
With pain in the heart I have to close this PR. Let's keep branch around.
